File: Codefun/alibaba/09033.cc
#include<bits/stdc++.h>
using namespace std;
#define int long long 
#define pb push_back 
#define MAXN 120000
int s[120000];
int Parent[MAXN];


map<int,int> mp;

void init(int n){
    for(int i=0;i<n;i++){
        Parent[i]=i;        
    }
}

// 并查集
int find(int x){
    if(Parent[x]==x)
        return x;
    else{
            Parent[x]=find(Parent[x]);   
            return Parent[x];
    }
}
void merge(int i,int j){
    Parent[find(j)] = find(i);
}

signed main() {
	int n;
	cin>>n;
	for(int i=1;i<=n;i++){
		cin>>s[i];
	}
	init(MAXN);
	int ans=0;
	for(int i=0;i<n-1;i++)
	{
		int lin,lin1;
		cin>>lin>>lin1;
		if(s[lin]==s[lin1])
		merge(lin,lin1);
	}
	for(int i=1;i<=n;i++){
		mp[find(i)]++;
	}
	int p=0;
	for(auto i : mp){
		if(s[i.first]==3){
			ans+=(i.second)*(i.second-1)/2*3;
		}
		else{
			ans+=(i.second)*(i.second-1)/2*5;
		}
		p+=15*(n-i.second)*(i.second);
	}
	cout<<ans+p/2<<endl; 
	return 0;
}

File: Codefun/alibaba/09033.py
import sys

sys.setrecursionlimit(100000)
n = int(input())
a = list(map(int, input().split()))

pa = list(range(n))


def find(x):
    return x if pa[x] == x else find(pa[x])


def merge(x: int, y: int):
    pa[find(y)] = find(x)


for _ in range(n - 1):
    u, v = map(int, input().split())
    u, v = u - 1, v - 1
    if a[u] == a[v]:
        merge(u, v)

d = dict()
for i in range(n):
    d[find(i)] = d.get(find(i), 0) + 1

p = 0
ans = 0
for k, v in d.items():
    if a[k] == 3:
        ans += v * (v - 1) // 2 * 3
    else:
        ans += v * (v - 1) // 2 * 5
    p += 15 * (n - v) * v

print(ans + p // 2)
